File: plant_care/lib/utils/care_sections.dart
/// Care guidance travels from the analyzer as a structured object, gets
/// flattened into a `Label: value` line per section (see
/// `_composeCareTipsFromCareMap`), and is stored on the plant as a single
/// `aiCareTips` string. This module owns that format: the label vocabulary used
/// to write it and the parser used to read it back.
///
/// Labels are baked in using the language that was active when the plant was
/// created, so the reader matches against *every* supported language rather
/// than the current locale — otherwise switching languages orphans the sections
/// of plants that already exist.
library;

/// Canonical, language-independent section keys.
class CareSection {
  const CareSection._();

  static const cultivar = 'cultivar';
  static const generalDescription = 'generalDescription';
  static const soil = 'soil';
  static const soilMoisture = 'soilMoisture';
  static const moistureCheck = 'moistureCheck';
  static const water = 'water';
  static const light = 'light';
  static const temperature = 'temperature';
  static const fertilizer = 'fertilizer';
  static const growthRate = 'growthRate';
  static const toxicity = 'toxicity';
  static const placement = 'placement';
  static const personality = 'personality';

  /// Write order for [composeCareTips].
  static const all = <String>[
    cultivar,
    generalDescription,
    soil,
    soilMoisture,
    moistureCheck,
    water,
    light,
    temperature,
    fertilizer,
    growthRate,
    toxicity,
    placement,
    personality,
  ];
}

/// Compact labels the analyzer returns alongside the prose sections, under
/// `care_recommendations.details`. These fill the key-value strips in the care
/// sheets and the one-line value on each care row — values the UI used to guess
/// with regexes over prose written in the user's language, which only ever
/// worked in English.
class CareDetail {
  const CareDetail._();

  static const wateringSeason = 'watering_season';
  static const lightHours = 'light_hours';
  static const lightType = 'light_type';
  static const temperatureOptimal = 'temperature_optimal';
  static const temperatureMinimum = 'temperature_minimum';
  static const fertilizerFrequency = 'fertilizer_frequency';
  static const fertilizerDose = 'fertilizer_dose';
  static const soilShort = 'soil_short';
  static const temperatureShort = 'temperature_short';
  static const fertilizerShort = 'fertilizer_short';
  static const placementShort = 'placement_short';

  static const all = <String>[
    wateringSeason,
    lightHours,
    lightType,
    temperatureOptimal,
    temperatureMinimum,
    fertilizerFrequency,
    fertilizerDose,
    soilShort,
    temperatureShort,
    fertilizerShort,
    placementShort,
  ];
}

/// These are labels for narrow cells; the prompt asks for 30 characters. Give
/// the model room to overshoot, but treat a paragraph as a failed instruction
/// and drop it so the UI falls back rather than blowing up its layout.
const _maxDetailLength = 64;

/// Reads `care_recommendations.details` into a clean string map, keeping only
/// keys the UI knows and values that are plausibly labels.
Map<String, String>? extractCareDetails(dynamic careRecommendations) {
  if (careRecommendations is! Map) return null;
  final raw = careRecommendations['details'];
  if (raw is! Map) return null;

  final out = <String, String>{};
  for (final key in CareDetail.all) {
    final value = raw[key];
    if (value == null || value is Map || value is List) continue;
    final s = value.toString().trim();
    if (s.isEmpty || s.length > _maxDetailLength) continue;
    out[key] = s;
  }
  return out.isEmpty ? null : out;
}

/// Section labels used when writing the blob, in [lang].
Map<String, String> careLabelsForLang(String lang) {
  switch (lang) {
    case 'ru':
      return const {
        'cultivar': 'Культивар', 'generalDescription': 'Общее описание',
        'soil': 'Почва', 'soilMoisture': 'Влажность почвы',
        'moistureCheck': 'Проверка влажности', 'water': 'Полив',
        'light': 'Освещение', 'temperature': 'Температура',
        'fertilizer': 'Удобрения', 'growthRate': 'Скорость роста',
        'toxicity': 'Токсичность', 'placement': 'Размещение',
        'personality': 'Характер',
      };
    case 'uk':
      return const {
        'cultivar': 'Культивар', 'generalDescription': 'Загальний опис',
        'soil': 'Ґрунт', 'soilMoisture': 'Вологість ґрунту',
        'moistureCheck': 'Перевірка вологості', 'water': 'Полив',
        'light': 'Освітлення', 'temperature': 'Температура',
        'fertilizer': 'Добрива', 'growthRate': 'Швидкість росту',
        'toxicity': 'Токсичність', 'placement': 'Розміщення',
        'personality': 'Характер',
      };
    case 'de':
      return const {
        'cultivar': 'Kultivar', 'generalDescription': 'Allgemeine Beschreibung',
        'soil': 'Erde', 'soilMoisture': 'Bodenfeuchtigkeit',
        'moistureCheck': 'Feuchtigkeitsprüfung', 'water': 'Wasser',
        'light': 'Licht', 'temperature': 'Temperatur',
        'fertilizer': 'Dünger', 'growthRate': 'Wachstumsrate',
        'toxicity': 'Toxizität', 'placement': 'Standort',
        'personality': 'Charakter',
      };
    case 'es':
      return const {
        'cultivar': 'Cultivar', 'generalDescription': 'Descripción general',
        'soil': 'Suelo', 'soilMoisture': 'Humedad del suelo',
        'moistureCheck': 'Verificación de humedad', 'water': 'Agua',
        'light': 'Luz', 'temperature': 'Temperatura',
        'fertilizer': 'Fertilizante', 'growthRate': 'Tasa de crecimiento',
        'toxicity': 'Toxicidad', 'placement': 'Ubicación',
        'personality': 'Personalidad',
      };
    case 'fr':
      return const {
        'cultivar': 'Cultivar', 'generalDescription': 'Description générale',
        'soil': 'Sol', 'soilMoisture': 'Humidité du sol',
        'moistureCheck': "Vérification de l'humidité", 'water': 'Eau',
        'light': 'Lumière', 'temperature': 'Température',
        'fertilizer': 'Engrais', 'growthRate': 'Taux de croissance',
        'toxicity': 'Toxicité', 'placement': 'Emplacement',
        'personality': 'Personnalité',
      };
    default:
      return const {
        'cultivar': 'Cultivar', 'generalDescription': 'General Description',
        'soil': 'Soil', 'soilMoisture': 'Soil Moisture',
        'moistureCheck': 'Moisture Check', 'water': 'Water',
        'light': 'Light', 'temperature': 'Temperature',
        'fertilizer': 'Fertilizer', 'growthRate': 'Growth Rate',
        'toxicity': 'Toxicity', 'placement': 'Placement',
        'personality': 'Personality',
      };
  }
}

/// Every label the writer has ever emitted, in any supported language, mapped
/// to its canonical key. Includes the analyzer's own English spellings, which
/// differ slightly from the app's (`Moisture` vs `Soil Moisture`).
const _labelToKey = <String, String>{
  // English (app + analyzer)
  'cultivar': CareSection.cultivar,
  'name': CareSection.cultivar,
  'general description': CareSection.generalDescription,
  'soil': CareSection.soil,
  'soil moisture': CareSection.soilMoisture,
  'moisture': CareSection.soilMoisture,
  'moisture check': CareSection.moistureCheck,
  'water': CareSection.water,
  'watering': CareSection.water,
  'light': CareSection.light,
  'temperature': CareSection.temperature,
  'fertilizer': CareSection.fertilizer,
  'growth rate': CareSection.growthRate,
  'growth': CareSection.growthRate,
  'toxicity': CareSection.toxicity,
  'placement': CareSection.placement,
  'personality': CareSection.personality,
  // Russian
  'культивар': CareSection.cultivar,
  'общее описание': CareSection.generalDescription,
  'почва': CareSection.soil,
  'влажность почвы': CareSection.soilMoisture,
  'проверка влажности': CareSection.moistureCheck,
  'полив': CareSection.water,
  'освещение': CareSection.light,
  'температура': CareSection.temperature,
  'удобрения': CareSection.fertilizer,
  'скорость роста': CareSection.growthRate,
  'токсичность': CareSection.toxicity,
  'размещение': CareSection.placement,
  'характер': CareSection.personality,
  // Ukrainian
  'загальний опис': CareSection.generalDescription,
  'ґрунт': CareSection.soil,
  'вологість ґрунту': CareSection.soilMoisture,
  'перевірка вологості': CareSection.moistureCheck,
  'освітлення': CareSection.light,
  'добрива': CareSection.fertilizer,
  'швидкість росту': CareSection.growthRate,
  'токсичність': CareSection.toxicity,
  'розміщення': CareSection.placement,
  // German
  'kultivar': CareSection.cultivar,
  'allgemeine beschreibung': CareSection.generalDescription,
  'erde': CareSection.soil,
  'bodenfeuchtigkeit': CareSection.soilMoisture,
  'feuchtigkeitsprüfung': CareSection.moistureCheck,
  'wasser': CareSection.water,
  'licht': CareSection.light,
  'temperatur': CareSection.temperature,
  'dünger': CareSection.fertilizer,
  'wachstumsrate': CareSection.growthRate,
  'toxizität': CareSection.toxicity,
  'standort': CareSection.placement,
  'charakter': CareSection.personality,
  // Spanish
  'descripción general': CareSection.generalDescription,
  'suelo': CareSection.soil,
  'humedad del suelo': CareSection.soilMoisture,
  'verificación de humedad': CareSection.moistureCheck,
  'agua': CareSection.water,
  'luz': CareSection.light,
  'temperatura': CareSection.temperature,
  'fertilizante': CareSection.fertilizer,
  'tasa de crecimiento': CareSection.growthRate,
  'toxicidad': CareSection.toxicity,
  'ubicación': CareSection.placement,
  'personalidad': CareSection.personality,
  // French
  'description générale': CareSection.generalDescription,
  'sol': CareSection.soil,
  'humidité du sol': CareSection.soilMoisture,
  "vérification de l'humidité": CareSection.moistureCheck,
  'eau': CareSection.water,
  'lumière': CareSection.light,
  'température': CareSection.temperature,
  'engrais': CareSection.fertilizer,
  'taux de croissance': CareSection.growthRate,
  'toxicité': CareSection.toxicity,
  'emplacement': CareSection.placement,
  'personnalité': CareSection.personality,
};

/// Resolves one written label to its canonical key, or null if unrecognised.
String? careLabelToKey(String raw) =>
    _labelToKey[_normalizeLabel(raw)];

/// Longest real label is ~26 chars; anything longer is prose, not a heading.
const _maxLabelLength = 40;

final _labelNoise = RegExp(r'^[\s>#*_\-•–—]+|[\s*_:]+$');

/// A bold label writes its closing `**` after the colon, so the body inherits it.
final _leadingEmphasis = RegExp(r'^[\s*_]+');

String _normalizeLabel(String raw) =>
    raw.replaceAll(_labelNoise, '').toLowerCase().trim();

/// Splits [blob] into canonical sections.
///
/// Recognises both shapes the writers produce: `Label: body` on a single line,
/// and a bare `Label` heading followed by its body. Unlabelled lines continue
/// the section above them, so multi-paragraph bodies survive intact. A section
/// ends where the next recognised label begins — nothing bleeds across.
Map<String, String> parseCareSections(String? blob) {
  if (blob == null || blob.trim().isEmpty) return const {};

  final buffers = <String, List<String>>{};
  String? current;

  void append(String text) {
    final section = current;
    if (section == null) return;
    buffers.putIfAbsent(section, () => <String>[]).add(text);
  }

  for (final rawLine in blob.split('\n')) {
    final line = rawLine.trim();
    if (line.isEmpty) {
      append('');
      continue;
    }

    final colon = line.indexOf(':');
    String? key;
    String remainder = '';

    if (colon > 0 && colon <= _maxLabelLength) {
      key = careLabelToKey(line.substring(0, colon));
      if (key != null) {
        remainder =
            line.substring(colon + 1).replaceFirst(_leadingEmphasis, '').trim();
      }
    }
    // Markdown-style heading on its own line, body follows.
    if (key == null && line.length <= _maxLabelLength) {
      key = careLabelToKey(line);
    }

    if (key != null) {
      current = key;
      buffers.putIfAbsent(key, () => <String>[]);
      if (remainder.isNotEmpty) append(remainder);
    } else {
      append(line);
    }
  }

  final out = <String, String>{};
  buffers.forEach((key, lines) {
    final body = lines.join('\n').trim();
    if (body.isNotEmpty) out[key] = body;
  });
  return out;
}

/// Writes [sections] back out in [lang], mirroring the format [parseCareSections]
/// reads. Keys absent from [sections] are skipped.
String? composeCareTips(Map<String, String?> sections, String lang) {
  final labels = careLabelsForLang(lang);
  final lines = <String>[];
  for (final key in CareSection.all) {
    final value = sections[key]?.trim();
    if (value == null || value.isEmpty) continue;
    lines.add('${labels[key]}: $value');
  }
  return lines.isEmpty ? null : lines.join('\n');
}

File: plant_care/functions/care-sections.js
/**
 * The care plan, as the server reads it.
 *
 * A plant's care guidance is stored as one flat `aiCareTips` string: a
 * `Label: body` line per section, written in whatever language was active when
 * the plant was added. The app owns the same format in lib/utils/care_sections.dart
 * — this is the reader half of it, ported so the assistant can quote the exact
 * paragraph the user is looking at instead of paraphrasing from memory.
 *
 * The two must stay in step. If a label is added on the app side and not here,
 * the section it introduces silently stops reaching the model: the parser will
 * treat the heading as body text and glue it onto whatever came before.
 */

// ── Canonical, language-independent section keys ────────────────────
const CARE_SECTION = {
  cultivar: 'cultivar',
  generalDescription: 'generalDescription',
  soil: 'soil',
  soilMoisture: 'soilMoisture',
  moistureCheck: 'moistureCheck',
  water: 'water',
  light: 'light',
  temperature: 'temperature',
  fertilizer: 'fertilizer',
  growthRate: 'growthRate',
  toxicity: 'toxicity',
  placement: 'placement',
  personality: 'personality',
};

/**
 * Every label either writer has ever emitted, in any supported language.
 *
 * Matching covers all six languages rather than the request locale on purpose:
 * the label was frozen into the blob when the plant was created, so a user who
 * has since switched the app to English still has a plant whose sections are
 * headed "Полив". Narrowing this to the current locale orphans them.
 */
const LABEL_TO_KEY = {
  // English (app + analyzer — the analyzer says "Moisture", the app "Soil Moisture")
  'cultivar': CARE_SECTION.cultivar,
  'name': CARE_SECTION.cultivar,
  'general description': CARE_SECTION.generalDescription,
  'soil': CARE_SECTION.soil,
  'soil moisture': CARE_SECTION.soilMoisture,
  'moisture': CARE_SECTION.soilMoisture,
  'moisture check': CARE_SECTION.moistureCheck,
  'water': CARE_SECTION.water,
  'watering': CARE_SECTION.water,
  'light': CARE_SECTION.light,
  'temperature': CARE_SECTION.temperature,
  'fertilizer': CARE_SECTION.fertilizer,
  'growth rate': CARE_SECTION.growthRate,
  'growth': CARE_SECTION.growthRate,
  'toxicity': CARE_SECTION.toxicity,
  'placement': CARE_SECTION.placement,
  'personality': CARE_SECTION.personality,
  // Russian
  'культивар': CARE_SECTION.cultivar,
  'общее описание': CARE_SECTION.generalDescription,
  'почва': CARE_SECTION.soil,
  'влажность почвы': CARE_SECTION.soilMoisture,
  'проверка влажности': CARE_SECTION.moistureCheck,
  'полив': CARE_SECTION.water,
  'освещение': CARE_SECTION.light,
  'температура': CARE_SECTION.temperature,
  'удобрения': CARE_SECTION.fertilizer,
  'скорость роста': CARE_SECTION.growthRate,
  'токсичность': CARE_SECTION.toxicity,
  'размещение': CARE_SECTION.placement,
  'характер': CARE_SECTION.personality,
  // Ukrainian
  'загальний опис': CARE_SECTION.generalDescription,
  'ґрунт': CARE_SECTION.soil,
  'вологість ґрунту': CARE_SECTION.soilMoisture,
  'перевірка вологості': CARE_SECTION.moistureCheck,
  'освітлення': CARE_SECTION.light,
  'добрива': CARE_SECTION.fertilizer,
  'швидкість росту': CARE_SECTION.growthRate,
  'токсичність': CARE_SECTION.toxicity,
  'розміщення': CARE_SECTION.placement,
  // German
  'kultivar': CARE_SECTION.cultivar,
  'allgemeine beschreibung': CARE_SECTION.generalDescription,
  'erde': CARE_SECTION.soil,
  'bodenfeuchtigkeit': CARE_SECTION.soilMoisture,
  'feuchtigkeitsprüfung': CARE_SECTION.moistureCheck,
  'wasser': CARE_SECTION.water,
  'licht': CARE_SECTION.light,
  'temperatur': CARE_SECTION.temperature,
  'dünger': CARE_SECTION.fertilizer,
  'wachstumsrate': CARE_SECTION.growthRate,
  'toxizität': CARE_SECTION.toxicity,
  'standort': CARE_SECTION.placement,
  'charakter': CARE_SECTION.personality,
  // Spanish
  'descripción general': CARE_SECTION.generalDescription,
  'suelo': CARE_SECTION.soil,
  'humedad del suelo': CARE_SECTION.soilMoisture,
  'verificación de humedad': CARE_SECTION.moistureCheck,
  'agua': CARE_SECTION.water,
  'luz': CARE_SECTION.light,
  'temperatura': CARE_SECTION.temperature,
  'fertilizante': CARE_SECTION.fertilizer,
  'tasa de crecimiento': CARE_SECTION.growthRate,
  'toxicidad': CARE_SECTION.toxicity,
  'ubicación': CARE_SECTION.placement,
  'personalidad': CARE_SECTION.personality,
  // French
  'description générale': CARE_SECTION.generalDescription,
  'sol': CARE_SECTION.soil,
  'humidité du sol': CARE_SECTION.soilMoisture,
  "vérification de l'humidité": CARE_SECTION.moistureCheck,
  'eau': CARE_SECTION.water,
  'lumière': CARE_SECTION.light,
  'température': CARE_SECTION.temperature,
  'engrais': CARE_SECTION.fertilizer,
  'taux de croissance': CARE_SECTION.growthRate,
  'toxicité': CARE_SECTION.toxicity,
  'emplacement': CARE_SECTION.placement,
  'personnalité': CARE_SECTION.personality,
};

// Longest real label is ~26 chars; past that it is prose, not a heading.
const MAX_LABEL_LENGTH = 40;

const LABEL_NOISE = /^[\s>#*_\-•–—]+|[\s*_:]+$/g;
// A bold label closes its `**` after the colon, so the body inherits it.
const LEADING_EMPHASIS = /^[\s*_]+/;

function normalizeLabel(raw) {
  return String(raw).replace(LABEL_NOISE, '').toLowerCase().trim();
}

/** Resolves one written label to its canonical key, or null if unrecognised. */
function careLabelToKey(raw) {
  return LABEL_TO_KEY[normalizeLabel(raw)] || null;
}

/**
 * Splits the stored blob into canonical sections.
 *
 * Recognises both shapes the writers produce: `Label: body` on one line, and a
 * bare `Label` heading with the body beneath. Unlabelled lines continue the
 * section above them, so multi-paragraph bodies survive. A section ends where
 * the next recognised label begins — nothing bleeds across.
 */
function parseCareSections(blob) {
  if (!blob || !String(blob).trim()) return {};

  const buffers = {};
  let current = null;

  const append = (text) => {
    if (!current) return;
    (buffers[current] = buffers[current] || []).push(text);
  };

  for (const rawLine of String(blob).split('\n')) {
    const line = rawLine.trim();
    if (!line) {
      append('');
      continue;
    }

    const colon = line.indexOf(':');
    let key = null;
    let remainder = '';

    if (colon > 0 && colon <= MAX_LABEL_LENGTH) {
      key = careLabelToKey(line.slice(0, colon));
      if (key) remainder = line.slice(colon + 1).replace(LEADING_EMPHASIS, '').trim();
    }
    // Markdown-style heading on its own line, body follows.
    if (!key && line.length <= MAX_LABEL_LENGTH) key = careLabelToKey(line);

    if (key) {
      current = key;
      buffers[key] = buffers[key] || [];
      if (remainder) append(remainder);
    } else {
      append(line);
    }
  }

  const out = {};
  for (const [key, lines] of Object.entries(buffers)) {
    const body = lines.join('\n').trim();
    if (body) out[key] = body;
  }
  return out;
}

// ── Topics ──────────────────────────────────────────────────────────

/**
 * The seven chat topics, and what each one is about in care-plan terms.
 *
 * A topic is a decision the owner makes about the plant, not the screen they
 * arrived from — which is why the "Soil Moisture" and "Placement" cards do not
 * get topics of their own. Grouping follows how the care sheets already compose
 * their bodies: the watering sheet shows `water` plus `moistureCheck`, so the
 * watering topic carries both.
 *
 * `diagnostics` and `general` deliberately map to nothing. A diagnosis is
 * grounded in health checks rather than in the standing plan, and the general
 * chat has no section to be reading.
 */
const TOPIC_SECTIONS = {
  water: [CARE_SECTION.water, CARE_SECTION.moistureCheck],
  soil: [CARE_SECTION.soil, CARE_SECTION.soilMoisture],
  light: [CARE_SECTION.light, CARE_SECTION.placement],
  temperature: [CARE_SECTION.temperature],
  fertilizer: [CARE_SECTION.fertilizer],
  diagnostics: [],
  general: [],
};

/** The compact `careDetails` labels worth showing alongside each topic. */
const TOPIC_DETAILS = {
  water: ['watering_season'],
  soil: ['soil_short'],
  light: ['light_hours', 'light_type', 'placement_short'],
  temperature: ['temperature_optimal', 'temperature_minimum', 'temperature_short'],
  fertilizer: ['fertilizer_frequency', 'fertilizer_dose', 'fertilizer_short'],
  diagnostics: [],
  general: [],
};

const TOPICS = Object.keys(TOPIC_SECTIONS);

function isKnownTopic(topic) {
  return typeof topic === 'string' && Object.hasOwn(TOPIC_SECTIONS, topic);
}

/**
 * The plan text behind one topic, exactly as the owner sees it in the sheet.
 *
 * Returned verbatim rather than summarised: the whole point is that the
 * assistant continues the paragraph on screen instead of restating a generic
 * version of it, and a paraphrase here would reintroduce the contradiction it
 * is meant to remove. Returns null when there is nothing to show, so the caller
 * can omit the block rather than print an empty heading.
 */
function careBriefForTopic(topic, careTips, careDetails) {
  if (!isKnownTopic(topic)) return null;

  const sections = parseCareSections(careTips);
  const bodies = TOPIC_SECTIONS[topic]
    .map((key) => sections[key])
    .filter((body) => body && body.trim());

  const details = TOPIC_DETAILS[topic]
    .map((key) => (careDetails || {})[key])
    .filter((value) => value && String(value).trim());

  if (!bodies.length && !details.length) return null;

  const parts = [];
  if (bodies.length) parts.push(bodies.join('\n'));
  if (details.length) parts.push(details.join(' · '));
  return parts.join('\n');
}

/**
 * The whole standing plan as prose, or null when the plant has none.
 *
 * The chat gets one section because the owner is reading one section. A health
 * check gets all of them: it is judging the plant as a whole, and it is the
 * thing allowed to rewrite the plan, so it has to see what it would be
 * replacing.
 *
 * Identity and personality are left out — they say nothing about how the plant
 * is doing, and every line here is paid for on every scan.
 */
const PLAN_OMITTED = [
  CARE_SECTION.cultivar,
  CARE_SECTION.generalDescription,
  CARE_SECTION.personality,
  CARE_SECTION.toxicity,
];

function wholeCarePlan(carePlan) {
  if (!carePlan) return null;
  const sections = parseCareSections(carePlan.tips);
  const lines = Object.entries(sections)
    .filter(([key, body]) => !PLAN_OMITTED.includes(key) && body && body.trim())
    .map(([key, body]) => `${key}: ${body.trim()}`);

  const details = carePlan.details || {};
  const detailLine = Object.entries(details)
    .filter(([, value]) => value && String(value).trim())
    .map(([key, value]) => `${key}=${value}`)
    .join(', ');
  if (detailLine) lines.push(`details: ${detailLine}`);

  return lines.length ? lines.join('\n') : null;
}

module.exports = {
  CARE_SECTION,
  wholeCarePlan,
  TOPICS,
  TOPIC_SECTIONS,
  TOPIC_DETAILS,
  isKnownTopic,
  careLabelToKey,
  parseCareSections,
  careBriefForTopic,
};
